packages/sdk/src/reliability_monitor/sender.ts

```diff
@@ -0,0 +1,57 @@
+import type { Peer, PeerId } from "@libp2p/interface";
+import { CoreProtocolResult, PeerIdStr } from "@waku/interfaces";
+import { Logger } from "@waku/utils";
+
+const log = new Logger("sdk:sender:reliability_monitor");
+
+const DEFAULT_MAX_ATTEMPTS_BEFORE_RENEWAL = 3;
+
+export class SenderReliabilityMonitor {
+  private attempts: Map<PeerIdStr, number> = new Map();
+  private readonly maxAttemptsBeforeRenewal =
+    DEFAULT_MAX_ATTEMPTS_BEFORE_RENEWAL;
+
+  public constructor(private renewPeer: (peerId: PeerId) => Promise<Peer>) {}
+
+  public async attemptRetriesOrRenew(
+    peerId: PeerId,
+    protocolSend: () => Promise<CoreProtocolResult>
+  ): Promise<void> {
+    const peerIdStr = peerId.toString();
+    const currentAttempts = this.attempts.get(peerIdStr) || 0;
+    this.attempts.set(peerIdStr, currentAttempts + 1);
+
+    if (currentAttempts + 1 < this.maxAttemptsBeforeRenewal) {
+      try {
+        const result = await protocolSend();
+        if (result.success) {
+          log.info(`Successfully sent message after retry to ${peerIdStr}`);
+          this.attempts.delete(peerIdStr);
+        } else {
+          log.error(
+            `Failed to send message after retry to ${peerIdStr}: ${result.failure}`
+          );
+          await this.attemptRetriesOrRenew(peerId, protocolSend);
+        }
+      } catch (error) {
+        log.error(
+          `Failed to send message after retry to ${peerIdStr}: ${error}`
+        );
+        await this.attemptRetriesOrRenew(peerId, protocolSend);
+      }
+    } else {
+      try {
+        const newPeer = await this.renewPeer(peerId);
+        log.info(
+          `Renewed peer ${peerId.toString()} to ${newPeer.id.toString()}`
+        );
+
+        this.attempts.delete(peerIdStr);
+        this.attempts.set(newPeer.id.toString(), 0);
+        await protocolSend();
```

Comment:
prev - https://github.com/waku-org/js-waku/pull/2130/files#r1758702483

here we call `.send` in both branches making no exit of the recursion (i.e it will be called over and over it seems to me)

so I think here we should do `.send` to `newPeer` instead because next time from here `attemptRetriesOrRenew` will be called for `peerIdStr` and at that point `this.attempts.` will not have it so it will just continue 

I believe this is the reason why you needed to implement `.stop` operation on the manager entity

Reply:
oh right, thanks for clarifying 
then it is not an issue - we can have a recursion here